git/github.go

```diff
@@ -0,0 +1,261 @@
+package git
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/bradleyfalzon/ghinstallation/v2"
+	"github.com/go-logr/logr"
+	"github.com/google/go-github/v45/github"
+	"github.com/redhat-appstudio/integration-service/helpers"
+	"golang.org/x/oauth2"
+)
+
+type GitHubAdapter struct {
+	GitProviderAdapter
+}
+
+// newGitHubAdapter creates a GitHubAdapter object
+func newGitHubAdapter(context context.Context, logger logr.Logger) *GitHubAdapter {
+	return &GitHubAdapter{
+		GitProviderAdapter{
+			context: context,
+			logger:  logger,
+		},
+	}
+}
+
+// getClient returns a GitHub client with authentication configured as well as the Application ID, if available.
+func (a *GitHubAdapter) getClient(gtpr *GitTriggeredPipelineRun) (*github.Client, int64, error) {
+	if gtpr.InstallationID > 0 {
+		var err error
+
+		ghAppIDBytes, found := gtpr.PaCSecret.Data["github-application-id"]
+		if !found {
+			return nil, 0, errors.New("failed to find github-application-id secret key")
+		}
+
+		ghAppID, err := strconv.ParseInt(string(ghAppIDBytes), 10, 64)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		ghPrivateKey, found := gtpr.PaCSecret.Data["github-private-key"]
+		if !found {
+			return nil, 0, errors.New("failed to find github-private-key secret key")
+		}
+
+		transport, err := ghinstallation.NewAppsTransport(http.DefaultTransport, ghAppID, ghPrivateKey)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		client := github.NewClient(&http.Client{Transport: transport})
+		installToken, _, err := client.Apps.CreateInstallationToken(
+			context.Background(),
+			gtpr.InstallationID,
+			&github.InstallationTokenOptions{},
+		)
+
+		if err != nil {
+			return nil, 0, err
+		}
+
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: installToken.GetToken()},
+		)
+
+		oauthClient := oauth2.NewClient(context.Background(), ts)
+		return github.NewClient(oauthClient), ghAppID, nil
+	}
+
+	return nil, 0, errors.New("building a GitHub client without an Installation ID is not yet supported")
+}
+
+// reportState reports PipelineRun status back to GitHub pull requests using the CheckRuns API.
+func (a *GitHubAdapter) reportState(gtpr *GitTriggeredPipelineRun) error {
+	if gtpr.EventType != "pull_request" {
+		return nil
+	}
+
+	client, appID, err := a.getClient(gtpr)
+	if err != nil {
+		return err
+	}
+
+	_, err = a.createOrUpdateCheckRun(client, appID, gtpr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// createOrUpdateCheckRun reports state using GitHub's CheckRun API. It can only be used with GitHub Apps.
+func (a *GitHubAdapter) createOrUpdateCheckRun(client *github.Client, appID int64, gtpr *GitTriggeredPipelineRun) (*github.CheckRun, error) {
+	var cr *github.CheckRun
+
+	var status, title, summary, conclusion string
+	var startTime time.Time
+	var completionTime time.Time
+
+	checkName := AppName + " / " + gtpr.Component + " / " + gtpr.TestScenario
```

Comment:
It doesn't need to be unique. GitHub assumes the latest check run using a given name should supersede prior runs. It will only display the newer run too.

Reply:
Sorry, the two commits I meant are two PRs. Each PR in Github should get a pipelinerun result, right?

Reply:
Sort of. A build `PipelineRun` will be kicked off for the head commit of the PR's source branch which, upon successful completion, may trigger multiple integration `PipelineRuns` (one per `IntegrationTestScenario`). There can be multiple `CheckRuns` for a given commit and so the name is used to group them under distinct workflows. Each `CheckRun` also has a unique `ExternalID` which matches the name of the `PipelineRun`. There should only ever be a single `CheckRun` per `PipelineRun`. Does that help?

Reply:
Yes, it helps to understand. Thanks